File: content/examples/other/leetcode/problems/kth-smallest-element-in-a-bst/example-1/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function kthSmallestInorderList(root, k) {
  const values = [];
  (function inorder(node) {
    if (!node) return;
    inorder(node.left);
    values.push(node.val);
    inorder(node.right);
  })(root);
  return values[k - 1];
}

if (require.main === module) {
  const root = new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4));
  console.log(kthSmallestInorderList(root, 1));
}

module.exports = { TreeNode, kthSmallestInorderList };

File: content/examples/other/leetcode/problems/kth-smallest-element-in-a-bst/example-1/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
  static class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
      this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
    }
  }

  static int kthSmallestInorderList(TreeNode root, int k) {
    List<Integer> values = new ArrayList<>();
    inorder(root, values);
    return values.get(k - 1);
  }

  static void inorder(TreeNode node, List<Integer> values) {
    if (node == null) return;
    inorder(node.left, values);
    values.add(node.val);
    inorder(node.right, values);
  }

  public static void main(String[] args) {
    TreeNode root = new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4));
    System.out.println(kthSmallestInorderList(root, 1));
  }
}
